File: Kaskus/GridViewPaging/GridViewPagingCs/Forms/FormMain.cs
namespace Dede.App.GridViewPaging.Forms
{
    using System;
    using System.Collections;
    using System.Windows.Forms;

    using Dede.App.GridViewPaging.Datas;
    using Dede.App.GridViewPaging.Pages;

    public partial class FormMain 
        : Form
    {
        private const Int32 PageSize = 10;
        private const Int32 BeginCurrentPage = 0;

        private RepositoryFactory repositoryFactory;

        public FormMain()
        {
            this.InitializeComponent();

            this.repositoryFactory = new RepositoryFactory(new RepositoryItemGrid());
            
            this.bindingNavigatorMain.BindingSource = this.bindingSourceMain;
            this.InitDataGridView();
            this.InitComboDataSource();
        }

        private void BindingSourceCurrentChanged(
            Object sender,
            EventArgs e)
        {
            this.BindDataGridView();
        }

        private void InitComboDataSource()
        {
            this.comboBoxDataSource.Items.Add(
                new DictionaryEntry
                    {
                        Key = typeof(RepositoryItemGrid).Name.Replace(@"Repository", String.Empty), 
                        Value = new RepositoryItemGrid()
                    });

            this.comboBoxDataSource.Items.Add(
                new DictionaryEntry
                {
                    Key = typeof(RepositoryDataTable).Name.Replace(@"Repository", String.Empty),
                    Value = new RepositoryDataTable()
                });

            this.comboBoxDataSource.DisplayMember = @"Key";
            this.comboBoxDataSource.ValueMember = @"Value";

            this.comboBoxDataSource.SelectedIndexChanged += (s, e) =>
                {
                    this.repositoryFactory =
                        new RepositoryFactory(
                            (IDataSourceRepository)((DictionaryEntry)this.comboBoxDataSource.SelectedItem).Value);

                    this.BindDataGridView();
                };

            this.comboBoxDataSource.SelectedIndex = 0;
        }

        private void InitDataGridView()
        {
            this.dataGridViewMain.Columns.Clear();
            this.dataGridViewMain.AllowUserToAddRows = false;
            this.dataGridViewMain.AutoGenerateColumns = false;
            this.dataGridViewMain.Columns.Add(new DataGridViewTextBoxColumn { DataPropertyName = "Id", Visible = false });
            this.dataGridViewMain.Columns.Add(new DataGridViewTextBoxColumn { DataPropertyName = "ItemString", HeaderText = @"String" });
            this.dataGridViewMain.Columns.Add(new DataGridViewTextBoxColumn { DataPropertyName = "ItemInt32", HeaderText = @"Int32" });
            this.dataGridViewMain.Columns.Add(new DataGridViewCheckBoxColumn { DataPropertyName = "ItemBoolean", HeaderText = @"Boolean" });
            this.dataGridViewMain.Columns.Add(new DataGridViewTextBoxColumn { DataPropertyName = "ItemDecimal", HeaderText = @"Decimal" });
            this.dataGridViewMain.Columns.Add(new DataGridViewTextBoxColumn { DataPropertyName = "ItemDateTime", HeaderText = @"DateTime" });
        }

        private void BindDataGridView()
        {
            var pageItem = this.bindingSourceMain.Current as PageList.ItemPage;
            var pageList = this.bindingSourceMain.DataSource as PageList;

            var dataResult =
                pageItem == null
                    ? this.repositoryFactory.GetData(PageSize, BeginCurrentPage)
                    : this.repositoryFactory.GetData(pageItem.PageSize, pageItem.CurrentPage);
            
            if ((pageList != null &&
                    pageList.TotalRecord != dataResult.TotalRecord) ||
                (pageItem == null))
            {
                this.bindingSourceMain.CurrentChanged -= this.BindingSourceCurrentChanged;
                this.bindingSourceMain.DataSource = new PageList(PageSize, dataResult.TotalRecord);
                this.bindingSourceMain.CurrentChanged += this.BindingSourceCurrentChanged;
            }

            this.dataGridViewMain.DataSource = dataResult.GetList();
        }
    }
}

File: Kaskus/GridViewPaging/GridViewPagingVb/Forms/FormMain.vb
Imports Dede.App.GridViewPaging.Datas
Imports Dede.App.GridViewPaging.Pages

Namespace Forms

    Public Class FormMain

        Private Const PageSize As Int32 = 10
        Private Const BeginCurrentPage As Int32 = 0

        Private _repositoryFactory As RepositoryFactory

        Public Sub New()
            InitializeComponent()

            _repositoryFactory = New RepositoryFactory(New RepositoryItemGrid())

            bindingNavigatorMain.BindingSource = bindingSourceMain
            InitDataGridView()
            InitComboDataSource()
        End Sub

        Private Sub BindingSourceCurrentChanged(sender As [Object], e As EventArgs)
            BindDataGridView()
        End Sub

        Private Sub InitComboDataSource()
            With comboBoxDataSource
                .Items.Add(New DictionaryEntry() With { _
                              .Key = GetType(RepositoryItemGrid).Name.Replace("Repository", [String].Empty), _
                              .Value = New RepositoryItemGrid()})
                .Items.Add(New DictionaryEntry() With { _
                              .Key = GetType(RepositoryDataTable).Name.Replace("Repository", [String].Empty), _
                              .Value = New RepositoryDataTable()})

                .DisplayMember = "Key"
                .ValueMember = "Value"

                AddHandler .SelectedIndexChanged,
                    Sub(s, e)
                        _repositoryFactory = New RepositoryFactory(CType((CType(comboBoxDataSource.SelectedItem, DictionaryEntry)).Value, IDataSourceRepository))
                        BindDataGridView()
                    End Sub

                .SelectedIndex = 0
            End With
        End Sub

        Private Sub InitDataGridView()
            With dataGridViewMain
                .Columns.Clear()
                .AllowUserToAddRows = False
                .AutoGenerateColumns = False
                .Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "Id", .Visible = False})
                .Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "ItemString", .HeaderText = "String"})
                .Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "ItemInt32", .HeaderText = "Int32"})
                .Columns.Add(New DataGridViewCheckBoxColumn() With {.DataPropertyName = "ItemBoolean", .HeaderText = "Boolean"})
                .Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "ItemDecimal", .HeaderText = "Decimal"})
                .Columns.Add(New DataGridViewTextBoxColumn() With {.DataPropertyName = "ItemDateTime", .HeaderText = "DateTime"})
            End With
        End Sub

        Private Sub BindDataGridView()
            With bindingSourceMain
                Dim pageItem = TryCast(.Current, PageList.ItemPage)
                Dim pageList = TryCast(.DataSource, PageList)

                Dim dataResult = If(pageItem Is Nothing, _repositoryFactory.GetData(PageSize, BeginCurrentPage), _repositoryFactory.GetData(pageItem.PageSize, pageItem.CurrentPage))

                If (pageList IsNot Nothing AndAlso pageList.TotalRecord <> dataResult.TotalRecord) OrElse (pageItem Is Nothing) Then
                    RemoveHandler .CurrentChanged, AddressOf BindingSourceCurrentChanged
                    .DataSource = New PageList(PageSize, dataResult.TotalRecord)
                    AddHandler .CurrentChanged, AddressOf BindingSourceCurrentChanged
                End If

                dataGridViewMain.DataSource = dataResult.GetList()
            End With
        End Sub

    End Class
End Namespace